resources/package.rb

```diff
@@ -0,0 +1,74 @@
+#
+# Cookbook:: syslog_ng
+# Resource:: package
+#
+# Copyright:: 2020, Ben Hughes <bmhughes@bmhughes.co.uk>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the spcific language governing prmissions and
+# limitations under the License.
+
+include SyslogNg::Cookbook::PackageHelpers
+
+property :packages, [String, Array],
+          default: lazy { default_packages(repo_include: package_repository, repo_exclude: package_repository_exclude) },
+          coerce: proc { |p| p.is_a?(Array) ? p : [p] }
+
+property :packages_exclude, [String, Array],
+          coerce: proc { |p| p.is_a?(Array) ? p : [p] },
+          default: []
+
+property :package_repository, [String, Array],
+          coerce: proc { |p| p.is_a?(Array) ? p : [p] },
+          description: 'Install packages from a specific repository/repositories'
+
+property :package_repository_exclude, [String, Array],
+          coerce: proc { |p| p.is_a?(Array) ? p : [p] },
+          description: 'Repository/repositories to exclude when installing packages'
+
+action_class do
+  def do_package_action(action)
+    packages = []
+    ruby_block 'get packages' do
+      block do
+        packages = new_resource.packages
+        log "Found #{packages.count} packages to #{action}."
+        if new_resource.packages_exclude
+          log "Excluding packages: #{new_resource.packages_exclude.join(', ')}."
+          packages.delete_if { |package| new_resource.packages_exclude.include?(package) }
+          log "There are #{packages.count} packages to #{action} after exclusion."
+        end
+      end
+    end
```

Comment:
Why isn't this just a package install block?

Reply:
It's a helper method so it can be lazily evaluated as if you add a custom repo (i.e. a COPR one to get the latest version on CentOS) it won't pick up the packages on the first chef run as the repo isn't there during compile. I've tried multiple ways and that's the only way I can get it to work immediately, it also avoids adding an intermediary variable to hold the packages in case they are supplied by property vs just being the defaults.

The way it was before was to check if the `packages` property was set then assign that to a variable if not assign the defaults from the repo search method. I felt it was tidier to use `defaults` on the property rather than this as it seemed the cleanest way.

The reason it exists at all is because the packages build for syslog-ng differ between versions as the extra modules seem to come and go, hence the options to exclude/include repos as from experience package conflicts turn up when using out of distro tree packages usually.

Reply:
If it needs to be there after the repos have been setup, why not just order it after the repository setup?

Reply:
As I have to use ShellOut to query the repo to get the available package names it happens at compile when the repo isn't setup so will return nothing. I can run it in a `ruby_block` to execute it after repo creation but then you have to deal with the scoping of a variable between the block resource and the recipe and it makes dealing with either using or overriding the default less elegant (not that this is 100% either), the previous version did this and it wasn't great looking (but did work).

We could drop the functionality completely but then there's the packages between versions/distros to deal with that would be pushed to the implementer so it's probably not worth having an install resource at all in that case.